SCons/Util.py

```diff
@@ -1472,66 +1472,180 @@ def f(self, x, y):
     setattr(obj, name, method)
 
 
-if hasattr(hashlib, 'md5'):
-    md5 = True
+# Default hash function and format. SCons-internal.
+_hash_function = None
+_hash_format = None
 
-    def MD5signature(s):
-        """
-        Generate md5 signature of a string
 
-        :param s: either string or bytes. Normally should be bytes
-        :return: String of hex digits representing the signature
-        """
-        m = hashlib.md5()
+def get_hash_format():
+    """
+    Retrieves the hash format or None if not overridden. A return value of None
+    does not guarantee that MD5 is being used; instead, it means that the
+    default precedence order documented in SCons.Util.set_hash_format is
+    respected.
+    """
+    return _hash_format
 
-        try:
-            m.update(to_bytes(s))
-        except TypeError as e:
-            m.update(to_bytes(str(s)))
 
-        return m.hexdigest()
+def set_hash_format(hash_format):
+    """
+    Sets the default hash format used by SCons. If hash_format is None or
+    an empty string, the default is determined by this function.
 
-    def MD5filesignature(fname, chunksize=65536):
-        """
-        Generate the md5 signature of a file
+    Currently the default behavior is to use the first available format of
+    the following options: MD5, SHA1, SHA256.
+    """
+    global _hash_format, _hash_function
+
+    _hash_format = hash_format
+    if hash_format:
+        hash_format_lower = hash_format.lower()
+        allowed_hash_formats = ['md5', 'sha1', 'sha256']
```

Comment:
Should this be: hashlib.algorithms_available or hashlib.algorithms_guaranteed?

Reply:
Thanks for the review! My understanding is that I can't use either of those because on FIPS they may advertise support for MD5 when in fact there is none. In that case, `hashlib.md5()` is simply not there.

The first version of my diff was much more open-ended up what it supported, basically allowing anything that could be called in `hashlib`. I can't remember if I changed it based on previous review feedback or a desire to have a 1-to-1 mapping between "supported" and "tested" algorithms. But I think the list as-is represents a good solution that is both supported and tested. We can always add more algorithms in the future, but in doing so we can then also confirm that the tests are updated to include that algorithm too.